### ISSUE TITLE
Application request when we haven't used client functionality lately. Optimistically trying directory fetches again.

### ISSUE THREAD

**Issue by maximilianosinski** [joelverhagen/TorSharp#96]
Tor doesn't connect, this is the last output from the logging:

`Jan 07 20:30:29.000 [notice] Application request when we haven't used client functionality lately. Optimistically trying directory fetches again.`

What can i do?

**Comment by joelverhagen** [comment]
This is a Tor log I've never seen. I recommend taking a look at the Tor documentation or Tor stack exchange site.

**Comment by maximilianosinski** [comment]
The first log that appears is another error:
```
Jan 07 22:42:19.000 [warn] Failed to open GEOIP file C:\Users\maxim\AppData\Local\Temp\Knapcode.TorSharp\ExtractedTools\tor-win64-12.0.1\data\tor\geoip.
Jan 07 22:42:19.000 [warn] Failed to open GEOIP file C:\Users\maxim\AppData\Local\Temp\Knapcode.TorSharp\ExtractedTools\tor-win64-12.0.1\data\tor\geoip6
```
although i can access these files

any solutions?

Edit: the path is wrong, the actual path is: 
```
C:\Users\maxim\AppData\Local\Temp\Knapcode.TorSharp\ExtractedTools\tor-win64-12.0.1\data\geoip
C:\Users\maxim\AppData\Local\Temp\Knapcode.TorSharp\ExtractedTools\tor-win64-12.0.1\data\geoip6
```

is there a way to change this with your library, or will you push a fix?

**Comment by joelverhagen** [comment]
@maximilianosinski, I can't reproduce this issue. It definitely does seem like a TorSharp bug as you describe, since the path is wrong. Could you share three things:
1. The full Tor output log.
2. A zip of the `C:\Users\maxim\AppData\Local\Temp\Knapcode.TorSharp` directory.
3. A minimal repro app, or at least what you initialize `TorSharpSettings` to be.

Somehow TorSharp thinks the Tor `DataDirectory` is the wrong thing.

**Comment by maximilianosinski** [comment]
As it looks, this error comes only if I specify ExitNodes, without ExitNodes everything works, but if I specify ExitNodes & StrictNodes = true not.

```
Jan 08 18:19:33.301 [notice] Tor can't help you if you use it wrong! Learn how to be safe at https://support.torproject.org/faq/staying-anonymous/
Jan 08 18:19:33.301 [notice] Tor 0.4.7.12 (git-f43a74975ed27d78) running on Windows 8 [or later] with Libevent 2.1.12-stable, OpenSSL 1.1.1s, Zlib 1.2.13, Liblzma N/A, Libzstd N/A and Unknown N/A as libc.
Jan 08 18:19:33.314 [notice] Read configuration file "C:\Users\maxim\AppData\Local\Temp\Knapcode.TorSharp\ExtractedTools\tor-win64-12.0.1\data\tor\torrc".
Jan 08 18:19:33.315 [warn] ControlPort is open, but no authentication method has been configured.  This means that any program on your computer can reconfigure your Tor.  That's bad!  You should upgrade your Tor controller as so
on as possible.
Jan 08 18:19:33.316 [notice] Opening Socks listener on 127.0.0.1:19050
Jan 08 18:19:33.316 [notice] Opening Control listener on 127.0.0.1:19051
Jan 08 18:19:33.316 [notice] Opened Socks listener connection (ready) on 127.0.0.1:19050
Jan 08 18:19:33.316 [notice] Opened Control listener connection (ready) on 127.0.0.1:19051
Jan 08 18:19:33.000 [warn] We've been configured to use (or avoid) nodes in certain countries, and we need GEOIP information to figure out which ones they are.
Jan 08 18:19:33.000 [warn] Failed to open GEOIP file C:\Users\maxim\AppData\Local\Temp\Knapcode.TorSharp\ExtractedTools\tor-win64-12.0.1\data\tor\geoip.
Jan 08 18:19:33.000 [warn] Failed to open GEOIP file C:\Users\maxim\AppData\Local\Temp\Knapcode.TorSharp\ExtractedTools\tor-win64-12.0.1\data\tor\geoip6.
Jan 08 18:19:33.000 [warn] We've been configured to use (or avoid) nodes in certain countries, and we need GEOIP information to figure out which ones they are.
Jan 08 18:19:33.000 [notice] Bootstrapped 0% (starting): Starting
Jan 08 18:19:33.000 [notice] Starting with guard context "default"
Jan 08 18:19:34.000 [notice] Bootstrapped 5% (conn): Connecting to a relay
Jan 08 18:19:55.000 [notice] Bootstrapped 10% (conn_done): Connected to a relay
Jan 08 18:19:55.000 [notice] Bootstrapped 14% (handshake): Handshaking with a relay
Jan 08 18:19:55.000 [notice] Bootstrapped 15% (handshake_done): Handshake with a relay done
Jan 08 18:19:55.000 [notice] Bootstrapped 20% (onehop_create): Establishing an encrypted directory connection
Jan 08 18:19:55.000 [notice] Bootstrapped 25% (requesting_status): Asking for networkstatus consensus
Jan 08 18:19:55.000 [notice] Bootstrapped 30% (loading_status): Loading networkstatus consensus
Jan 08 18:19:56.000 [notice] I learned some more directory information, but not enough to build a circuit: We have no usable consensus.
Jan 08 18:19:56.000 [notice] Bootstrapped 40% (loading_keys): Loading authority key certs
Jan 08 18:19:56.000 [notice] The current consensus has no exit nodes. Tor can only build internal paths, such as paths to onion services.
Jan 08 18:19:56.000 [notice] Bootstrapped 45% (requesting_descriptors): Asking for relay descriptors
Jan 08 18:19:56.000 [notice] I learned some more directory information, but not enough to build a circuit: We need more microdescriptors: we have 0/6184, and can only build 0% of likely paths. (We have 0% of guards bw, 0% of mid
point bw, and 0% of end bw (no exits in consensus, using mid) = 0% of path bw.)
Jan 08 18:19:56.000 [notice] I learned some more directory information, but not enough to build a circuit: We need more microdescriptors: we have 0/6184, and can only build 0% of likely paths. (We have 0% of guards bw, 0% of mid
point bw, and 0% of end bw (no exits in consensus, using mid) = 0% of path bw.)
Jan 08 18:19:56.000 [notice] Bootstrapped 50% (loading_descriptors): Loading relay descriptors
Jan 08 18:19:56.000 [notice] The current consensus contains exit nodes. Tor can build exit and internal paths.
```
ZIP: https://www.mediafire.com/file/nfo5nh9cwczkog1/Knapcode.TorSharp.zip/file

Code:

```
            var settings = new TorSharpSettings
            {
                PrivoxySettings = { Disable = true },
                TorSettings =
                {
                    ExitNodes = "{de}",
                    StrictNodes = true
                }
            };
            using (var httpClient = new HttpClient())
            {
                var fetcher = new TorSharpToolFetcher(settings, httpClient);
                await fetcher.FetchAsync();
            }
            using (var proxy = new TorSharpProxy(settings))
            {
                await proxy.ConfigureAndStartAsync();

                var handler = new HttpClientHandler
                {
                    Proxy = new WebProxy(new Uri("socks5://localhost:" + settings.TorSettings.SocksPort))
                };

                using (handler)
                using (var httpClient = new HttpClient(handler))
                {
                    var result = await httpClient.GetStringAsync("https://check.torproject.org/api/ip");

                    Console.WriteLine();
                    Console.WriteLine("Are we using Tor?");
                    Console.WriteLine(result);
                }

                proxy.Stop();
            }

**Comment by joelverhagen** [comment]
Thanks for the repro. I found the bug. A fix is coming!

**Comment by joelverhagen** [comment]
This is fixed in version 2.12.0.
